src/python/pants/core/goals/package.py

```diff
@@ -178,5 +185,35 @@ async def package_asset(workspace: Workspace, dist_dir: DistDir) -> Package:
     return Package(exit_code=0)
 
 
+@dataclass(frozen=True)
+class TraverseIfNotPackageTarget(ShouldTraverseDepsPredicate):
+    package_field_set_types: FrozenOrderedSet[PackageFieldSet]
+    roots: FrozenOrderedSet[Address]
+    always_traverse_roots: bool = True  # traverse roots even if they are package targets
+
+    def __init__(
+        self,
+        *,
+        union_membership: UnionMembership,
+        roots: Iterable[Address],
+        always_traverse_roots: bool = True,
+    ) -> None:
+        object.__setattr__(self, "package_field_set_types", union_membership.get(PackageFieldSet))
+        object.__setattr__(self, "roots", FrozenOrderedSet(roots))
+        object.__setattr__(self, "always_traverse_roots", always_traverse_roots)
+        super().__init__()
+
+    def __call__(self, target: Target, field: Dependencies | SpecialCasedDependencies) -> bool:
+        if isinstance(field, SpecialCasedDependencies):
+            return False
+        if self.always_traverse_roots and target.address in self.roots:
+            return True
+        for field_set_type in self.package_field_set_types:
+            if field_set_type.is_applicable(target):
+                # False means do not traverse dependencies of this target
+                return False
+        return True
```

Comment:
Alternatively you could use `any(field_set_type.is_applicable(target) for field_set_type in self.package_field_set_types)`

I tend to prefer `any` and `all` because they express intent better (e.g. "return if any is packageable" is the English way, and it uses the word `any`)

Reply:
Addressed in #19387